clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/AbstractStoreMetadata.java

```diff
@@ -250,4 +257,35 @@ public <K> void routeRequest(RequestContext requestContext, RecordSerializer<K>
     throw new VeniceClientException("Unknown request type: " + requestType);
   }
 
+  private void maybeLogNoReplicaErrorDetails(
+      int partitionId,
+      int version,
+      long requestId,
+      Set<String> excludedReplicas) {
+    String errorPrefix = String.format(NO_REPLICA_ERROR_PREFIX, storeName, partitionId);
+    if (REDUNDANT_LOGGING_FILTER.isRedundantException(errorPrefix)) {
+      // We logged the details for this store partition recently already
+      return;
+    }
+    List<String> replicas = getReplicas(version, partitionId);
+    Set<String> blockedReplicas = new HashSet<>();
+    Set<String> unhealthyReplicas = new HashSet<>();
+    for (String replica: replicas) {
+      if (instanceHealthMonitor.isInstanceBlocked(replica)) {
+        blockedReplicas.add(replica);
+      }
+      if (!instanceHealthMonitor.isInstanceHealthy(replica)) {
+        unhealthyReplicas.add(replica);
+      }
```

Comment:
This logging function may not catch the issue since the instance status can keep changing and maybe some replica was treated as unhealthy replica in the previous `getReplica`, but become healthy when logging.

One thing we can do is to extract the healthy replica detection logic from the `AbstractClientRoutingStrategy#getReplicas` and put it in `AbstractStoreMetadata`, so we will keep track of the unhealthy/blocked replicas, so that we can log them via RedundantExceptionFilter.

Reply:
Good idea, let me do that.